gb-apu/src/apu.rs

```diff
@@ -0,0 +1,179 @@
+use gb_bus::{Address, Bus, Error, FileOperation, IORegArea};
+use gb_clock::{Tick, Ticker};
+use sdl2::audio::AudioQueue;
+use std::{cell::RefCell, rc::Rc};
+
+use crate::{
+    channel::sound_channel::SoundChannel, control::frame_sequencer::FrameSequencer, ChannelType,
+    BUFFER_SIZE,
+};
+
+pub struct Apu {
+    cycle_counter: u16,
+    enabled: bool,
+    audio_queue: Rc<RefCell<AudioQueue<f32>>>,
+    buffer: [f32; BUFFER_SIZE],
+    buffer_i: usize,
+    sound_channels: Vec<SoundChannel>,
+    frame_sequencer: FrameSequencer,
+}
+
+impl Apu {
+    pub fn new(audio_queue: Rc<RefCell<AudioQueue<f32>>>) -> Apu {
+        // Channels order in vector is important !
+        let sound_channels = vec![
+            SoundChannel::new(ChannelType::SquareWave, true),
+            SoundChannel::new(ChannelType::SquareWave, false),
+            SoundChannel::new(ChannelType::WaveForm, false),
+            SoundChannel::new(ChannelType::Noise, false),
+        ];
+        Self {
+            cycle_counter: 0,
+            enabled: false,
+            audio_queue,
+            buffer: [0.0; BUFFER_SIZE],
+            buffer_i: 0,
+            sound_channels,
+            frame_sequencer: FrameSequencer::default(),
+        }
+    }
+    fn add_sample(&mut self) {
+        let (left_sample, right_sample) = self.mix();
+        self.buffer[self.buffer_i] = left_sample;
+        self.buffer_i += 1;
+        self.buffer[self.buffer_i] = right_sample;
+        self.buffer_i += 1;
+    }
+
+    fn mix(&self) -> (f32, f32) {
```

Comment:
this function can do stereo ? (responsible for the left/right side of the sound)

Reply:
It will eventually after I implement the left/right control register

Reply:
thank you for the clarification